### ISSUE TITLE
Disable "Does this place allow smoking?" quest in the UK (GB)

### ISSUE THREAD

**Issue by TurnrDev** [streetcomplete/StreetComplete#4518]
Smoking indoors is illegal in the UK, so this quest is rather spammy as all indoor places would automatically be "No".

**Comment by TurnrDev** [comment]
I'm happy to make this change, but would this be accepted?

**Comment by Helium314** [comment]
I had also asked for this (#3856), but the quest will not be disabled on a per-country base because it's disabled by default.

**Comment by matkoniecz** [comment]
> It is possible in the future to disable the quest completely for certain countries where we know 100% that smoking is not allowed anywhere in public places.

https://github.com/streetcomplete/StreetComplete/issues/3856#issuecomment-1064629471

It can be disabled by default and in addition blocked in some specific countries.

> indoor places would automatically be "No".

What about outdoor ones?

**Comment by mnalis** [comment]
My reasoning (as the quest author) was that it was **_already disabled_** by default for everybody, and that _people who do not want to be spammed_ should/would not _manually enable_ something that they know would spam them.

Just like if there are (for example) no special Kosher places in the country, people would not go manually enable Kosher quest (which is disabled by default) and than afterwards complain that Kosher quest (which they willingly and manually enabled) is spamming them. (I mean, it's not like once manually enabled they can never turn it off again! If you expended effort to turn it on against the default; you can also expend effort to turn it off once you find out you do not actually want it)

> It can be disabled by default and in addition blocked in some specific countries.

Technically it can. But especially since the **_default already does what people expect_** (i.e. quest disabled meaning no-spamming), I find it _**bad idea to needlessly take the choice away** from the people_ (e.g. **forbid** them from temporarily enabling the quest if they do find there is some rare specialized Kosher restaurant in some neighborhood - especially if people take the time to manually temporarily enable some quest because they feel that such rare answers are worth much more than if they were common)

TL;DR - Just don't go manually enabling quests you don't want to answer, eh?

**Comment by Helium314** [comment]
> TL;DR - Just don't go manually enabling quests you don't want to answer, eh?

This is still the same absurd argumentation as before. You assume people are either not travelling at all, or should create per-country presets because of a single quest.
StreetComplete _is_ already taking choice from people, e.g. by limiting the (disabled by default) cash quest to specific countries. Or by mostly not asking the (disabled by default) roof shape quest if roofs are usually flat in a country.

**Comment by Helium314** [comment]
Actually I think there is no need to disabled it entirely in UK, as the spammy part can be avoided with country-dependent element selection (only ask if `outdoor_seating != no` in UK).

**Comment by TurnrDev** [comment]
That does seem like a good solution! There's even a map of which countries ban smoking indoors, which we can use: https://en.wikipedia.org/wiki/List_of_smoking_bans

**Comment by mnalis** [comment]
> That does seem like a good solution! There's even a map of which countries ban smoking indoors, which we can use: https://en.wikipedia.org/wiki/List_of_smoking_bans

That wiki page in specific, however, I would heavily **advise against**. I'd much prefer someone on the ground who also
has knowledge / has researched the situation (like @TurnrDev in UK), or preferably few of them (e.g. after local
communication group discussion?) to open an issue.

For example, for my little part of the Europe I semi-regularly visit, there is at least `75%` of incorrect data there (i.e. wikipedia mismatching reality). And even for parts that are factually correct, it is extremely hard to parse (i.e. it requires a _huge_ amount of work to get from that wiki to a  table of `country|smoking_inside_allowed|smoking_outside_allowed` values) .

**Comment by matkoniecz** [comment]
> (only ask if outdoor_seating and outdoor_seating != no in UK).

That would require adding extra feature or splitting into separate quests, both are problematic.

**Comment by mnalis** [comment]
> That would require adding extra feature or splitting into separate quests, both are problematic.

That is true, especially splitting the quest into two seems like a very bad idea to me. 

But, extra feature / function like `and isQuestInCountry(UK,SE)` which could be used inside `elementFilter` might conceivably be used in other quests as needed (provided someone volunteers to write it and it is deemed acceptable / not kludgy, of course). 

AFAIK, SC  does not currently have anything like that, though (we only have _operators_ and _tags_ in `elementFilter` currently, right?). Is so, perhaps instead of a "function", a virtual tag like `__QUEST_IS_IN_COUNTRY` (which has value like `UK` or `DE` etc.) might be invented, and then normal operators might be used on it, e.g.:

```
elementFilter = """
  amenity ~ nightclub|stripclub 
  or amenity ~ restaurant|pub|bar and  (outdoor_seating != no or __QUEST_IS_IN_COUNTRY !~ UK|SE)
  and (!smoking or smoking older today -8 years) 
 [...]
``` 

It might lead to quite complex `elementFilter` values though (example above is just a small _grossly oversimplified **chunk**_ of smoking quest filter, to illustrate the possible usage if something like that were to be implemented)

**Comment by mnalis** [comment]
Devoting some more time to investigate the wikipedia articles on the matter, I have a question for @TurnrDev (and other UK citizens).  https://en.wikipedia.org/wiki/Smoking_ban_in_England#Reaction_to_the_ban says:

> In 2012, it was reported that "Five years after the introduction of the smoking ban in England, almost seven out of 10 licensees want the legislation amended to allow for separate smoking rooms in pubs".

Unfortunately, wikipedia article does not reference newer statistics since 2012. Do you have links to reference **newer statistics**? Because, if the information from 2012. citing `70%` of pubs wanting `smoking=separate` has not changed drastically, it seems quite a good reason to retain the quest in England at least for `smoking=outside` and `smoking=separate` answers (i.e. making **only** the `smoking=yes` answer irrelevant).

**Comment by mnalis** [comment]
Summary of all options so far
---
(for those not wanting to read whole thread, so it can be linked to. I'll update it if I see other options become available, and/or people mention me to update it):

- allow to **check for country of the quest in elementFilter** so special filters could be implemented for `UK`. (it needs new functionality to be written as prerequisite, though: https://github.com/streetcomplete/StreetComplete/issues/4528)
- **split the Smoking quest into two quests**, one for `UK` (and similar countries if later reported), and one for the rest of the world. (it has obvious maintenance increase implications, though, and opens untidy precedent for other quests to follow. And quest selector would confuse users too by having two smoking quests)
- **do nothing**, if _"disabled by default"_ is deemed good enough not to be spammy to vast majority (it may make people who manually enable it and travel often and do not wish to use presets feel spammed, though. This is still my personal favorite, because):
  - there is about `43%` of `smoking=outside`, so that answer at least should remain always
  - nightclubs and similar pay-to-enter venues seem to have `smoking=separate`, so that answer should remain there at least.
  - somewhat old (but still latest) wikipedia articles mention that `~70%` of `amenity=pub` also wanted/had `smoking=separate` in England, so unless that changed drastically, quest with possible `smoking=separate` answer should be shown there too (even when they don't qualify for `smoking=yes`).
  - the quest is **already** _"disabled by default"_, and I do not feel that **many** people who enabled it **manually** travel so **frequently**  that creating/switching presets would undully overtax them (especially as other points seem to indicate that the quest should still be asked in many places in `UK` which seems to have **banned _only_** the `smoking=yes` answer)  .
  - parties interested in change should go read not only this issue, but the whole of discussions in https://github.com/streetcomplete/StreetComplete/pull/3784, https://github.com/streetcomplete/StreetComplete/issues/539, https://github.com/streetcomplete/StreetComplete/pull/3865, https://github.com/streetcomplete/StreetComplete/issues/3856 and linked pages. Many things have been considered and tuned for and much research and other efforts expended to reach current solution. It is not perfect, but it is quite complex already and reasons are subtle. Updating the logic would need revisiting all of those concerns brought up previously, in order not to break it.
  - keeping the _status quo_ requires no work at all :smile: 

- **disable the smoking quest unconditionally in UK** for all users, even if they want to enable this disabled_by_default quest (Statistics show that only about `52%` of the OSM tags are `smoking=no` there, and all other allow some sort of smoking. This is the one I personally am heavily against, as it seems mapping `smoking` tag is quite relevant in `UK` despite the _"no unrestricted smoking inside"_ ban )

**Comment by TurnrDev** [comment]
Under UK law, `smoking=seperated` is still prohibited, so I imagine any tags with that probably mean `smoking=outside` or are outdated information?

Also, do we implement `smoking=isolated` like the Wiki mentions or did we skip it for its low usage?

**Comment by TurnrDev** [comment]
In the meantime, I'll set up a challenge on MapRoulette to check the usage of anything other than `smoking=no` and `smoking=outside` in the UK

**Comment by Helium314** [comment]
How do only ask if there is outdoor seating: https://github.com/streetcomplete/StreetComplete/commit/25bb3f853344a2f1b77c4b5f69a991752989c90c
Just needed to change the quest to use OsmElementQuestType and copy some code from cycleway quest.
(this is untested, and will not compile due to QuestsModule being unchanged)

**Comment by mnalis** [comment]
> Under UK law, smoking=seperated is still prohibited, so I imagine any tags with that probably mean smoking=outside or are outdated information?
>
> Also, do we implement smoking=isolated like the Wiki mentions or did we skip it for its low usage?

@TurnrDev It was initially implemented, but then removed due to feedback that the users would be confused by only slight differences between `separated` and `isolated` (i.e. as nothing in practice is really `100%` isolated, even opening the doors of "fully isolated room" would allow some smoke to escape, so it is technically then all just various grades of `separated` and never _really_ `isolated` - as you won't have dual-door [airlock chamber](https://www.nasa.gov/centers/johnson/engineering/integrated_environments/human_space_environment_testing/iss_airlock_chamber/index.html) with vacuum pumps in most nightclubs and pubs).

So it was deemed that `isolated` was just a specific (more rigid) case of `separated` and that the differences are quite subjective  and gradual instead of a clear cut - so that `Only in separated areas` answer would support both (but always marking as more generic  / less precise `smoking=separated`). 
So you might treat both `separated` and `isolated` as the same thing - some area where smoking is allowed near some other area where smoking is not allowed, with wide range of possibility of smoke escape (0.1% to 99%) from one to the other.

So I guess when the mentioned example article https://www.vice.com/en/article/z3xwa4/this-is-what-englands-nightclub-smoking-areas-looked-like-last-night says:

> Anyway, for reasons related to a virus, we've not had access to nightclub smoking areas for quite some time now – or, at least, not constant access. It's been sporadic at best

that  _"sporadic access to **nightclub smoking areas** due to virus"_ was taking more about [smoking=isolated](https://wiki.openstreetmap.org/wiki/Key:smoking) (which it seems **_still exist_** in UK - or at least did a year ago), and the pictures there are showing them at `smoking=outside` (even if there is no outside _seating_, just standing in case of nightclubs and similar amenities)

But, as noted, SC does not distinguish between `isolated` and `separated`, so its `Only in separated areas` is intentionally designed to cover both cases.

**Comment by peternewman** [comment]
> So I guess when the mentioned example article https://www.vice.com/en/article/z3xwa4/this-is-what-englands-nightclub-smoking-areas-looked-like-last-night says:
> 
> > Anyway, for reasons related to a virus, we've not had access to nightclub smoking areas for quite some time now – or, at least, not constant access. It's been sporadic at best
> 
> that _"sporadic access to **nightclub smoking areas** due to virus"_ was taking more about [smoking=isolated](https://wiki.openstreetmap.org/wiki/Key:smoking) (which it seems **_still exist_** in UK - or at least did a year ago), and the pictures there are showing them at `smoking=outside` (even if there is no outside _seating_, just standing in case of nightclubs and similar amenities)

This is probably getting a bit OT, but no, it doesn't mean that @mnalis . In the UK, for large parts of the Coronavirus pandemic, nightclubs were one of a small number of types of businesses which were legally banned from operating, so with no nightclubs, there was no access to their outside smoking areas either (or at least as far as I'm aware no-one bothered to gather at them when large gatherings weren't also illegal).

As talk-gb mentioned, in nearly all cases we should be mapping it's `outside` or `no` due to the law!

**Comment by mnalis** [comment]
Yes, `talk-gb` seems clear so far. What I think could be done to improve the `smoking quest`:

- **remove all answers except `smoking=no` and `smoking=outside`** when quest is in `GB` (and possibly other counties in the future that discuss the subject in their local communication channel and submit the request), in order to limit users from submitting illegal answers (and reduce their possible confusion). 
  It would need code to access the country of the quest from [SmokingAllowedForm context](https://github.com/streetcomplete/StreetComplete/blob/00f00f60b192bda332dae5249b5fdcd099148f29/app/src/main/java/de/westnordost/streetcomplete/quests/smoking/SmokingAllowedForm.kt#L23-L27 
) (hopefully not too complex for someone who is more intimate with SC code than me)

 - when in `GB`, **limit the showing of the smoking quest** but only for amenities that are [likely](https://github.com/streetcomplete/StreetComplete/blob/00f00f60b192bda332dae5249b5fdcd099148f29/app/src/main/java/de/westnordost/streetcomplete/quests/smoking/SmokingAllowedForm.kt#L18-L21) to **not** have _"outside **standing** smoking areas"_ (**not to be confused** with _"outside **seating** areas"_!)

    So, in countries like `UK`, it would only show the smoking quest to `amenity=fast_food` only if it has `outdoor_seating=yes|...`, but it would show the quest to `amenity=nightclub` even if it has `outdoor_seating=no`.

    It would need  `elementFilter` parser support in readable and reusable form like suggested in https://github.com/streetcomplete/StreetComplete/issues/4528. If something like that is implemented, I'd be happy to update smoking quest `elementFilter` specific rules to accomodate that.

- If option(s) above are implemented, we also need to update the test code in [AddSmokingTest.kt](https://github.com/streetcomplete/StreetComplete/blob/master/app/src/test/java/de/westnordost/streetcomplete/quests/smoking/AddSmokingTest.kt) to cover UK-alike situations as well as existing ones; with ample comments. Luckily, the country support for automated test seems to [already be available](https://github.com/streetcomplete/StreetComplete/blob/00f00f60b192bda332dae5249b5fdcd099148f29/app/src/test/java/de/westnordost/streetcomplete/quests/cycleway/AddCyclewayTest.kt#L273-L285), so I could probably offer help there too.

**Comment by TurnrDev** [comment]
Sounds good and fair

**Comment by westnordost** [comment]
There might be a misunderstanding here. It is not possible in StreetComplete to enable a quest by default only in certain countries. There are two mechanisms available:

- Disable a quest _by default_ everywhere, i.e. user is able to turn it back on.
- Disable a quest completely for certain countries / only enable it in certain countries, i.e. users outside of the defined list of countries are not able to turn it back on again at all.

**Comment by wetneb** [comment]
Yes that's what I was suspecting (as mentioned above).
Would you consider having the smoking quest enabled by default if it were only available in white-listed countries where we have checked that it is useful?

**Comment by westnordost** [comment]
I am not sure... probably not... after all, e.g. the halal quest is also disabled by default but available in any country if the user really wants to survey that.
For smoking, e.g. in Germany, it is allowed only in a small fraction of establishments, still if someone wants to survey all that, the quest should not be not available to him at all.

Also, anyway, I guess oftentimes it is actually necessary to go inside to find out if smoking is allowed or not, and that is already one reason why the quest is disabled by default.
